packages/ethereum/contracts/stake/HoprStakeBase.sol

```diff
@@ -0,0 +1,532 @@
+  mapping(address => uint256) public redeemedFactorIndex; // The last index of redeemed boost factor factor of an account. It defines the length of the “redeemedFactor” mapping.
+  mapping(uint256 => bool) public isBlockedNft; // Type index of HoprBoost NFT that is not accepted in this season
+
+  mapping(address => Account) public accounts; // It stores the locked token amount, earned and claimed rewards per account.
+  uint256 public totalLocked; // Total amount of tokens being locked in the incentive program.
+  uint256 public availableReward; // Total amount of reward tokens currently available in the lock.
+
+  // setup ERC1820
+  IERC1820Registry private constant ERC1820_REGISTRY = IERC1820Registry(0x1820a4B7618BdE71Dce8cdc73aAB6C95905faD24);
+  bytes32 private constant TOKENS_RECIPIENT_INTERFACE_HASH = keccak256('ERC777TokensRecipient');
+
+  event Sync(address indexed account, uint256 indexed increment);
+  event Staked(address indexed account, uint256 indexed actualAmount);
+  event Released(address indexed account, uint256 indexed actualAmount);
+  event RewardFueled(uint256 indexed amount);
+  event Redeemed(address indexed account, uint256 indexed boostTokenId, bool indexed factorRegistered);
+  event Claimed(address indexed account, uint256 indexed rewardAmount);
+  event NftBlocked(uint256 indexed typeIndex);
+  event NftAllowed(uint256 indexed typeIndex);
+
+  /**
+   * @dev Provide basic parameters for the new staking season.
+   * Transfer owner role to the new owner address.
+   * At deployment, it also registers the staking contract as an ERC777 recipient.
+   * @param _newOwner address Address of the new owner. This new owner can reclaim any ERC20 and ERC721 token being accidentally sent to the lock contract.
+   * @param _programStart uint256 Timestamp from which the new staking season starts.
+   * @param _programEnd uint256 Timestamp till which the new staking season ends.
+   * @param _basicFactorNumerator uint256 Mumerator for the basic APY.
+   * @param _boostCap uint256 Cap for staked tokens to enjoy extra boost
+   * @param _nftAddress address Address of the NFT contract.
+   * @param _lockToken address Address of the stake token xHOPR.
+   * @param _rewardToken address Address of the reward token wxHOPR.
+   */
+  constructor(
+    address _newOwner,
+    uint256 _programStart,
+    uint256 _programEnd,
+    uint256 _basicFactorNumerator,
+    uint256 _boostCap,
+    address _nftAddress,
+    address _lockToken,
+    address _rewardToken
+  ) {
+    // set program parameters
+    PROGRAM_START = _programStart;
+    PROGRAM_END = _programEnd;
+    BASIC_FACTOR_NUMERATOR = _basicFactorNumerator;
+    BOOST_CAP = _boostCap;
+    transferOwnership(_newOwner);
+    ERC1820_REGISTRY.setInterfaceImplementer(address(this), TOKENS_RECIPIENT_INTERFACE_HASH, address(this));
+    // implement in favor of testing
+    uint256 chainId;
+    assembly {
+      chainId := chainid()
+    }
+    if (chainId != 100) {
```

Comment:
Should we remove the defaults and set the attributes all the time via the constructor? That way the logic becomes more clear.

Reply:
No. What you meant by the "generic" source code is very HOPR specific already which requires deployed token addresses and the "deployment information" is embedded in an inherited contract as described in the specifications. In reality, the "deployment information" is `[]`